### PR TITLE
feat: Add support for client-side prerequisite events.

### PR DESCRIPTION
**Pull request by kinyoklion**
BEGIN_COMMIT_OVERRIDE
feat: Add support for client-side prerequisite events.
feat: Add support for client-side visibility for all_flags_state.
END_COMMIT_OVERRIDE

SDK-803

### CODE REVIEW

File: priv/flags-segments-put-data.json
Comment (kinyoklion) [inline priv/flags-segments-put-data.json:1]: Updating this test data to have consistent clientSideAvailability with the clientSide field.

File: src/ldclient_eval.erl [hunk 1]
```diff
@@ -40,8 +40,8 @@ target_match
 }.
 
 -type all_flags_state_options() :: #{
-    with_reasons => boolean()
-%%    client_side_only => boolean(), % TODO: Support.
+    with_reasons => boolean(),
+    client_side_only => boolean()
```
Comment (kinyoklion) [inline src/ldclient_eval.erl:44]: It seemed reasonable to implement this at the same time, and that allows the contract tests to be passed as well.

File: test/ldclient_eval_SUITE.erl [hunk 2]
```diff
@@ -810,6 +812,101 @@ rule_match_rollout_not_in_experiment(_) ->
     ExpectedEvents = ActualEvents.
 
 all_flags_state(_) ->
```
Comment (kinyoklion) [inline test/ldclient_eval_SUITE.erl:814]: This test was supposed to be without reasons and the other test with reasons, but it originally had a mistake. So both tests were with reasons.

File: test/ldclient_eval_SUITE.erl [hunk 3]
```diff
@@ -967,170 +1064,41 @@ all_flags_state(_) ->
...
-        <<"target-me">> := true
-    } = ldclient_eval:all_flags_state(#{key => <<"userKeyA">>, kind => <<"user">>}, #{with_reasons => true}, default).
-
 all_flags_state_offline(_) ->
     #{
         <<"$flagsState">> := #{},
         <<"$valid">> := false
     } = ldclient_eval:all_flags_state(#{key => <<"userKeyA">>, kind => <<"user">>}, #{with_reasons => true}, offline).
 
+  all_flags_state_client_side_only(_) ->
```
Comment (kinyoklion) [inline test/ldclient_eval_SUITE.erl:1073]: New test for client-side visibility.

File: src/ldclient_eval.erl [hunk 4]
```diff
@@ -127,17 +147,26 @@ all_flags_state(_Context, _Options, _Tag, _, not_initialized) ->
 all_flags_state(Context, #{with_reasons := WithReason} = _Options, Tag, Offline, store_initialized) ->
     error_logger:warning_msg("Called allFlagsState before client initialization; using last known values from data store."),
     all_flags_state(Context, #{with_reasons := WithReason} = _Options, Tag, Offline, initialized);
-all_flags_state(Context, #{with_reasons := WithReason} = _Options, Tag, _, initialized) ->
+all_flags_state(Context, #{with_reasons := WithReason} = Options, Tag, _, initialized) ->
     FeatureStore = ldclient_config:get_value(Tag, feature_store),
-    AllFlags = [Flag || Flag = {_, FlagValue} <- FeatureStore:all(Tag, features), is_not_deleted(FlagValue)],
+    AllFlags = [Flag || Flag = {_, FlagValue} <- FeatureStore:all(Tag, features), is_not_deleted(FlagValue), is_visible(FlagValue, Options)],
```
Comment (kinyoklion) [inline src/ldclient_eval.erl:152]: Update the list comprehension to filter flags based on visibility.

File: src/ldclient_eval.erl [hunk 5]
```diff
@@ -113,6 +113,26 @@ all_flags_state(Context, Options, Tag) ->
 is_not_deleted(#{deleted := true}) -> false;
 is_not_deleted(_) -> true.
 
+-spec is_prereq_of(FlagKey :: ldclient_flag:key(), Event :: ldclient_event:event()) -> boolean().
+is_prereq_of(FlagKey, #{data := #{prereq_of := PrereqOf}} = _Event) ->
```
Comment (kinyoklion) [inline src/ldclient_eval.erl:117]: If prereq_of is set to the flag we are evaluating, then that flag was evaluated as a direct prerequisite.

**Review by keelerm84 — APPROVED** [review]

**Pull request merged**